components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/graphQL/GraphQLAPIHandler.java

```diff
@@ -0,0 +1,317 @@
+import org.apache.axis2.Constants;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpStatus;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.commons.json.JsonUtil;
+import org.apache.synapse.config.Entry;
+import org.apache.synapse.core.SynapseEnvironment;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.apache.synapse.core.axis2.Axis2Sender;
+import org.apache.synapse.rest.AbstractHandler;
+import org.apache.synapse.transport.nhttp.NhttpConstants;
+import org.apache.synapse.transport.passthru.util.RelayUtils;
+import org.wso2.carbon.apimgt.gateway.handlers.security.APISecurityConstants;
+import graphql.language.Definition;
+import graphql.language.Document;
+import graphql.language.Field;
+import graphql.language.OperationDefinition;
+import graphql.language.OperationDefinition.Operation;
+import graphql.language.Selection;
+import graphql.parser.Parser;
+import graphql.schema.GraphQLSchema;
+import graphql.schema.idl.SchemaParser;
+import graphql.schema.idl.TypeDefinitionRegistry;
+import graphql.schema.idl.UnExecutableSchemaGenerator;
+import graphql.validation.ValidationError;
+import graphql.validation.Validator;
+import org.wso2.carbon.apimgt.impl.APIConstants;
+
+import java.io.IOException;
+import java.util.Base64;
+import java.net.URLDecoder;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.xml.stream.XMLStreamException;
+
+public class GraphQLAPIHandler extends AbstractHandler {
+
+    private static final String QUERY_PATH_STRING = "/?query=";
+    private static final String GRAPHQL_SCOPE_ROLE_MAPPING = "GRAPHQL_SCOPE_ROLE_MAPPING";
+    private static final String GRAPHQL_SCOPE_OPERATION_MAPPING = "GRAPHQL_SCOPE_OPERATION_MAPPING";
+    private static final String REST_SUB_REQUEST_PATH = "REST_SUB_REQUEST_PATH";
+    private static final String API_TYPE = "API_TYPE";
+    private static final String GRAPHQL_API = "GRAPHQL";
+    private static final String UNICODE_TRANSFORMATION_FORMAT = "UTF-8";
+    private static final String INVALID_QUERY = "INVALID QUERY";
+    private static final String SCOPE_ROLE_MAPPING = "ScopeRoleMapping";
+    private static final String SCOPE_OPERATION_MAPPING = "ScopeOperationMapping";
+    private static final String GRAPHQL_IDENTIFIER = "_graphQL";
+    private static final String CLASS_NAME_AND_METHOD = "_GraphQLAPIHandler_handleRequest";
+    private static final Log log = LogFactory.getLog(GraphQLAPIHandler.class);
+    private static GraphQLSchema schema = null;
+    private static Validator validator;
+
+    public GraphQLAPIHandler() {
+        validator = new Validator();
+    }
+    private String apiUUID;
```

Comment:
shall we move this to the top where the other variable declarations are.

Reply:
Fixed